### ISSUE TITLE
Refactoring Android logcat to allow for publishing and subscribing

### ISSUE THREAD

**Issue by jeffherman** [google/mobly#486]
The [mobly.controller.android_device](https://github.com/google/mobly/blob/master/mobly/controllers/android_device.py) logcat functionality currently creates a file in `/tmp/logs` ([populated by the Mobly test runner](https://github.com/google/mobly/blob/master/mobly/logger.py#L135)).

While this allows for test artifact preservation, it doesn't allow the test to easily make use of the live logcat stream. A test could attach a Unix `tail -f` subprocess but due to an [issue](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/882147) with `inotify` on OverlayFS on Linux, it is often the case that any particular logcat line may be delayed and, therefore, cannot be used for realtime processing.

Another approach would be to create an entirely new logcat publisher class (a subclass of `threading.Thread`) and subscriber with `subscribe()` and `unsubscribe()` methods along with a logcat subscriber class (an abstract base class) that would handle new log lines. An example of a subscriber class would something that simply saves all raw logcat lines to a file (such as the current system) or something that's monitoring a particular regular expression and updating some state variables.

**Comment by xpconanfan** [comment]
I don't think we will change the current logcat mechanism to a different way.
We will likely preserve the current logcat and provide new APIs to make it easier to consume the content.
